data/tutorials/wf_03_metaprogramming.md

```diff
@@ -0,0 +1,454 @@
+---
+id: metaprogramming
+title: Preprocessors and PPXs
+description: |
+  An introduction to metaprogramming in OCaml, including preprocessors, PPX
+  extensions and the `ppxlib` library.
+category: "guides"
+date: 2022-07-25T21:07:30-00:00
+---
+
+
+# Preprocessors
+
+Preprocessors are programs meant to be called at compile time, so that they
+alter the program's source code before the actual compilation. They can be very
+useful for several things, such as the inclusion of a file, conditional
+compilation, or use of macros; but they're also useful in OCaml to generate a function from a
```

Comment:
I'm referring to this kind of macros https://gcc.gnu.org/onlinedocs/cpp/Macros.html (which is I think a fairly standard name, easy to google in case)

I speak about "local DSL" here: https://github.com/ocaml/ocaml.org/pull/535/files#diff-53ab2933d3906f740d8b4ead948ee84a719ff58dab72216d6147247d9468514aR331-R337
I'll try to see if it can be fit here, but on my opinion it is a bit early and OCaml-specific for an intro on preprocessors.

For the boilerplate code generation, I'll try to add something about it!

Reply:
> I'm referring to this kind of macros https://gcc.gnu.org/onlinedocs/cpp/Macros.html (which is I think a fairly standard name, easy to google in case)

Ok. That link describes the kind of macros you find in C. When saying macros, people also often have [the kind of macros in mind you find in Rust or Scheme](https://en.wikipedia.org/wiki/Macro_(computer_science)#Syntactic_macros), particularly in a functional programming context. Concretely in the context of OCaml, there are also discussions once in a while about introducing a concept called modular macros, which a few people see as a possible future alternative to PPX and others see it as a nice future addition. 

By the way, the other points you list ("the inclusion of a file" and "conditional compilation") describe the kind of things preprocessing can solve, whereas "macros" describe a certain kind of preprocessing ("macro" falls in the same category as "PPX" I'd say). I'd list more things preprocessing can solve instead of having this certain kind of preprocessing in that list, something along the lines:
> They can be very useful for several things, such as the inclusion of a file, conditional
compilation, boilerplate generation or extending the language with local DSLs.

What do you think? If you prefer to leave it as "the inclusion of a file, conditional compilation, or use of macros", I think it would be nice to make clear you're referring to C-macros (or, at least to make clear that this doc won't be about modular macros).

Reply:
Thanks for the info on macros! I did not know all of that. I think that the readers background can vary a lot, and a name clearly referring to something might be referring to something else for another person, or a "classic" example can be completely not classic for someone else

So, I like a lot your suggestion of not mentioning "other preprocessor" terminology such as macro (C or rust or any kind!). I like a lot your proposition, but I would simplify it by removing "with local DSLs", as I was often confused by what a DSL exactly is, and I'd like to keep this intro more oriented towards "preprocessor beginners".

Reply:
Sounds great!